File: packages/dotnet/src/Yourdudeken.Mpesa/Services/B2PochiService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Yourdudeken.Mpesa.Config;
using Yourdudeken.Mpesa.Http;
using Yourdudeken.Mpesa.Auth;

namespace Yourdudeken.Mpesa.Services
{
    public class B2PochiService
    {
        private readonly HttpClientWrapper _httpClient;
        private readonly AuthService _auth;
        private readonly MpesaConfig _config;
        private readonly string _baseUrl;

        public B2PochiService(HttpClientWrapper httpClient, AuthService auth, MpesaConfig config)
        {
            _httpClient = httpClient;
            _auth = auth;
            _config = config;
            _baseUrl = httpClient.GetBaseUrl();
        }

        public async Task<string> Send(string phonenumber, int amount, string remarks, string occasion = null,
            string resultUrl = null, string timeoutUrl = null, string shortCodeType = "B2C")
        {
            var url = $"{_baseUrl}/mpesa/b2pochi/v1/paymentrequest";
            var body = new Dictionary<string, object>
            {
                { "OriginatorConversationID", GetFormattedTimestamp() },
                { "InitiatorName", _config.InitiatorName },
                { "SecurityCredential", GenerateSecurityCredential() },
                { "CommandID", "BusinessPayToPochi" },
                { "Amount", amount },
                { "PartyA", _config.B2cShortcode },
                { "PartyB", PhoneValidator(phonenumber) },
                { "Remarks", remarks },
                { "Occasion", occasion ?? "" },
                { "ResultURL", resultUrl ?? _config.B2pochiResultUrl },
                { "QueueTimeOutURL", timeoutUrl ?? _config.B2pochiTimeoutUrl }
            };

            var token = await _auth.GetAccessToken(shortCodeType);
            return await _httpClient.Post(url, body, token);
        }

        private string PhoneValidator(string phoneNo)
        {
            if (phoneNo.StartsWith("+")) phoneNo = phoneNo.Substring(1);
            if (phoneNo.StartsWith("0")) phoneNo = "254" + phoneNo.Substring(1);
            else if (phoneNo.StartsWith("7")) phoneNo = "254" + phoneNo;
            return phoneNo;
        }

        private string GetFormattedTimestamp()
        {
            return DateTime.Now.ToString("yyyyMMddHHmmss");
        }

        private string GenerateSecurityCredential() => "";
    }
}

File: packages/go/mpesa/b2pochi_service.go
package mpesa

type B2PochiService struct {
	httpClient *HTTPClient
	auth       *Auth
	helpers    *Helpers
	baseURL    string
}

func NewB2PochiService(httpClient *HTTPClient, auth *Auth, config *Config) *B2PochiService {
	return &B2PochiService{
		httpClient: httpClient,
		auth:       auth,
		helpers:    NewHelpers(config),
		baseURL:    httpClient.GetBaseURL(),
	}
}

func (s *B2PochiService) Send(phonenumber string, amount int, remarks string, occasion string, resultURL string, timeoutURL string, shortCodeType string) (map[string]interface{}, error) {
	url := s.baseURL + "/mpesa/b2pochi/v1/paymentrequest"
	body := map[string]interface{}{
		"OriginatorConversationID": s.helpers.GetFormattedTimestamp(),
		"InitiatorName":            s.helpers.GetConfig("initiator_name"),
		"SecurityCredential":       "",
		"CommandID":                "BusinessPayToPochi",
		"Amount":                   amount,
		"PartyA":                   s.helpers.GetConfig("b2c_shortcode"),
		"PartyB":                   s.helpers.PhoneValidator(phonenumber),
		"Remarks":                  remarks,
		"Occasion":                 occasion,
		"ResultURL":                resultURL,
		"QueueTimeOutURL":          timeoutURL,
	}

	token, err := s.auth.GetAccessToken(shortCodeType)
	if err != nil {
		return nil, err
	}

	return s.httpClient.Post(url, body, token)
}
